### PR TITLE
Fixes issue affecting queries on Pointer arrays

### LINKED ISSUE #3169: Improper handling of query parameters depending on their order for relational arrays.
During investigation of iOS SDK bug https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/issues/1067 I've noticed that one of my queries fails when working on 32 bit device. At first, I thought it's due to recent SDK update from 1.13.0 to 1.14.2 but I was able to reproduce it with both SDK versions. After few hours of debugging how requests are made on 32/64 bit SDK I came to conclusion that ParseServer treats differently requests that queries fields with one direct object pointer vs array of those pointers. Because of the way how iOS SDK stores request parameters internally (in NSDictionary which is unordered collection) JSONs producend on 32 bit devices are different then the one from 64 bit devices (presented below) and requests on 32 bit devices is failing.

### Steps to reproduce

1. Ids known for two objects with type fooo 
```swift
    let fooId = "SLa1pur9QB"
    let fooId2 = "zDtZFtdx9X"
```
2. Couple of objects("GameScore") with relation array to "fooo" objects
```swift
let foo = PFObject(withoutDataWithClassName: "fooo", objectId: fooId)
let foo2 = PFObject(withoutDataWithClassName: "fooo", objectId: fooId2)
let gameScore = PFObject(className:"GameScore")
gameScore["foooVal"] = [foo, foo2]
```
3. Query to find objects of type GameScore where object foo2 is stored in "fooVal" array
```swift
let foo = PFObject(withoutDataWithClassName: "fooo", objectId: fooId)
let gameScoreQuery = PFQuery(className:"GameScore")
gameScoreQuery.whereKey("foooVal", equalTo: foo)
gameScoreQuery.findObjectsInBackground { (objects, error) in
    print("error \(error)")
    print("objects = \(objects?.count)")
}
```

I'll also present case where this doesn't matter due to fact that fields containg just one relation objects are stored differently by ParseServer

1. Save object of type GameScore2 with direct relation to fooo
```swift
let foo2 = PFObject(withoutDataWithClassName: "fooo", objectId: fooId2)
let gameScore = PFObject(className:"GameScore2")
gameScore["foooVal"] = foo2
```
2. Query
```swift
let foo2 = PFObject(withoutDataWithClassName: "fooo", objectId: fooId2)
let gameScoreQuery = PFQuery(className:"GameScore2")
gameScoreQuery.whereKey("foooVal", equalTo: foo2)
gameScoreQuery.findObjectsInBackground { (objects, error) in
    print("error \(error)")
    print("objects = \(objects?.count)")
}
```

#### Expected Results

Ordering of parameters in both queries should not influence results.

#### Actual Outcome

_1. Query on array field_

Produced JSON on 32 bit devices:
```
{"where":{"foooVal":{"objectId":"SLa1pur9QB","className":"fooo","__type":"Pointer"}},"_method":"GET"}
```
Produced JSON on 64 bit devices:
```
{"where":{"foooVal":{"__type":"Pointer","className":"fooo","objectId":"SLa1pur9QB"}},"_method":"GET"}
```

_Result:_
On 64 bit devices all values are returned correctly, on 32 bit empty array is returned(refer to logs section)

_2. Query on direct releation field_

32 bit device:
```
{"where":{"foooVal":{"objectId":"zDtZFtdx9X","className":"fooo","__type":"Pointer"}},"_method":"GET"}
```
64 bit device
```
{"where":{"foooVal":{"__type":"Pointer","className":"fooo","objectId":"zDtZFtdx9X"}},"_method":"GET"}
```

_Result:_
In both cases proper values is returned

### Environment Setup

- **Server**
  - parse-server version: 2.2.25-beta.1 commit: c8823f296c6128b991f69fba4e602c8d004a4658
  - Operating System:     64bit Amazon Linux 2016.09 v3.1.0 running Node.js
  - Hardware:             t2.micro instance
  - Localhost or remote server?: AWS Cloud

- **Database**
  - MongoDB version: 3.2.11
  - Hardware:        mLab Sandbox
  - Localhost or remote server?: mLab

### Logs/Trace

_1. query on relation array_
64 bit request/response log from server:
```
verbose: REQUEST for [GET] /parse/classes/GameScore: {
  "where": {
    "foooVal": {
      "__type": "Pointer",
      "className": "fooo",
      "objectId": "SLa1pur9QB"
    }
  }
}

verbose: RESPONSE from [GET] /parse/classes/GameScore: {
  "response": {
    "results": [
      SOME_PROPER_VALUES
    ]
  }
}
```

32 bit request/response log from server:
```
verbose: REQUEST for [GET] /parse/classes/GameScore: {
  "where": {
    "foooVal": {
      "objectId": "SLa1pur9QB",
      "className": "fooo",
      "__type": "Pointer"
    }
  }
}

verbose: RESPONSE from [GET] /parse/classes/GameScore: {
  "response": {
    "results": []
  }
} results=[]
```

_2. query on direct relation field_
64 bit request/response log from server
```
verbose: REQUEST for [GET] /parse/classes/GameScore2: {
  "where": {
    "foooVal": {
      "__type": "Pointer",
      "className": "fooo",
      "objectId": "zDtZFtdx9X"
    }
  }
}
verbose: RESPONSE from [GET] /parse/classes/GameScore2: {
  "response": {
    "results": [
      SOME_PROPER_VALUES
    ]
  }
}
```
32 bit request/response log from server:
```
verbose: REQUEST for [GET] /parse/classes/GameScore2: {
  "where": {
    "foooVal": {
      "objectId": "zDtZFtdx9X",
      "className": "fooo",
      "__type": "Pointer"
    }
  }
}
verbose: RESPONSE from [GET] /parse/classes/GameScore2: {
  "response": {
    "results": [
    SOME_PROPER_VALUES
    ]
  }
}
```
I guess that if field contains direct relation to object query is translated from above form to CLASS_NAME$OBJECT_ID before sending it to MongoDB, and in case of field containg array query from above is passed directly to MongoDB which is why it's returning 0 objects. So either such translation should happen in both cases or SDK should create queries in specific order (although I coulnd't find any statement in documentation for REST Api about such requirement) or maybe there's a way to convince MongoDB to not look at parameter order, just "content".

### PR DESCRIPTION
**Pull request by flovilmart**
Fixes #3169

### CODE REVIEW

File: spec/ParseQuery.spec.js [hunk 1]
```diff
@@ -2750,4 +2750,38 @@ describe('Parse.Query testing', () => {
...
+      const where = {
+        score: {
+          objectId: score1.id,
+          className: 'Score',
+          __type: 'Pointer',
+        }
+      }
+      return require('request-promise').post({
+        url: Parse.serverURL + "/classes/Game",
+        json: { where, "_method":"GET" },
```
Comment (acinader) [inline spec/ParseQuery.spec.js]: whitespace nit after :

File: spec/ParseQuery.spec.js [hunk 2]
```diff
@@ -2750,4 +2750,38 @@ describe('Parse.Query testing', () => {
...
+        headers: {
+          'X-Parse-Application-Id': Parse.applicationId,
+          'X-Parse-Javascript-Key': Parse.javaScriptKey
+        }
+      });
+    }).then((response) => {
+      expect(response.results.length).toBe(1);
+      done();
+    }, (err) => {
```
Comment (acinader) [inline spec/ParseQuery.spec.js]: nice terse fp shorthand for this is }, done.fail}; the error will get output

Comment (flovilmart) [inline spec/ParseQuery.spec.js]: Yep!

File: spec/ParseQuery.spec.js (hunk 1 above)
Comment (flovilmart) [inline spec/ParseQuery.spec.js]: 🙊

**Comment by bhirshman1** [comment]
Thank you for fixing this issue!!  My app is running on a fork of parse-community/parse-server-example, not parse-server.  Will this change also apply to the community version of parse-server-example?  I'm new to github and any words on how I can make use of your fix would be greatly appreciated.

**Comment by flovilmart** [comment]
This should be good, just make sure Parse-server dependency is at the right version, we'll release a new version soon.

**Comment by bhirshman1** [comment]
Awesome thanks!

**Pull request merged**